### PR TITLE
Very hacky trend view

### PR DESCRIPTION
**Pull request by achamayou**
I've been thinking about something like this for a bit, so here's an extremely hacked up implementation for the sake of the argument.
This doesn't scale particularly well to lots of metrics, not if we do it for every value. But it could be fine to do it only for the 4/6/8/whatever most changed values in a diff. And/or users could configure "pinned" metrics they want this detail for.

Also, and somewhat coincidentally, this doesn't suffer from the typical spacing/overwriting issues we have now, like #49. It probably gives a better sense of the stddev/zscore than alpha/width ever could too (#34).

### CODE REVIEW

**Comment by ghost** [comment]
sparks@[308 aka 20191224.3](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=308&view=results) vs master ewma over 20 builds from [166](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=166&view=results) to [304](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=304&view=results)
images

**Comment by letmaik** [comment]
Oooh I like it!

**Comment by ghost** [comment]
sparks@[309 aka 20191225.1](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=309&view=results) vs master ewma over 20 builds from [166](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=166&view=results) to [304](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=304&view=results)
images

**Comment by achamayou** [comment]
@letmaik great, I'll clean it up then. If we want to keep multiple views, one thing that I was thinking of is to use something in the commit message to decide which to render, eg. "Added buffering [trend]" would use the trend view rather than the snapshot.

**Comment by ghost** [comment]
sparks@[310 aka 20191225.2](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=310&view=results) vs master ewma over 20 builds from [166](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=166&view=results) to [304](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=304&view=results)
images

**Comment by letmaik** [comment]
Had a closer look at the plots. Two things:
- I find the scientific format in most cases less readable. 
- The only thing missing from the existing "simple" plots is the +/- percentage indication. Maybe this could be put into the titles somehow or in a second line (maybe `suptitle(..)`), like "↓ 10%" and "↑ 5%", maybe even colored green/red. With this I think the original plots wouldn't be needed anymore and could be removed at some point.

**Comment by achamayou** [comment]
Very much agree with your second point, just haven’t found a good way to do that yet. I think the other display is still somewhat more compact, but I’ve also tried three columns and it works fairly well too while taking less space.

Scientific format is less readable, I do agree, but it’s quite convenient because it’s super easy to give a constant amount of precision in a constant amount of space. At the moment we have sometimes long and awkward labels.
Could we maybe consider that the absolute numbers are secondary compared to the +/-% and so it’s ok for them to be less readable? Otherwise I will look for a way to put them at the top/bottom where horizontal space is less problematic I think.

**Comment by letmaik** [comment]
I assume you don't like it, but what about displaying in decimal format "if it fits"?

**Comment by achamayou** [comment]
Sounds very reasonable!

**Comment by ghost** [comment]
sparks@[311 aka 20191227.1](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=311&view=results) vs master ewma over 20 builds from [166](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=166&view=results) to [304](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=304&view=results)
images

**Comment by achamayou** [comment]
In fact it's slightly awkward, it's simple for the branch value, but the target value is an ewma, so it's typically a value with a lot of digits after the decimal point. So it needs to go something like:

1. render the branch value
2. work out the length
3. if it's short enough, find out how many digits after the decimal point it has
3a. if not, then use scientific format
4. and then render the target with the same precision

That only mostly works though, if there's a big change (order of magnitude or more), the rendering will be unsatisfactorily imprecise or overly precise.

**Comment by ghost** [comment]
sparks@[312 aka 20191228.1](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=312&view=results) vs master ewma over 20 builds from [166](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=166&view=results) to [304](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=304&view=results)
images

**Comment by achamayou** [comment]
It still beats defaulting to scientific all the time, I think.

**Comment by achamayou** [comment]
I wonder if we should use the config to let users define the beneficial direction for metrics and use a typical green/red scheme?

**Comment by ghost** [comment]
sparks@[313 aka 20191228.2](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=313&view=results) vs master ewma over 20 builds from [166](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=166&view=results) to [304](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=304&view=results)
images

**Comment by ghost** [comment]
sparks@[314 aka 20191228.3](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=314&view=results) vs master ewma over 20 builds from [166](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=166&view=results) to [304](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=304&view=results)
images

**Comment by achamayou** [comment]
One potential issue I hadn't thought of is that now the scales are different (based on the raw values rather than the normalised % delta), so 5% on one plot can be a long arrow, while being pretty short on another plot.

On the other hand, it's the width of the distribution over 2*span that decides the scale of the arrow, and that seems very appropriate (ie. if you get wild swings on a metric, a 10% change there is much less meaningful than a 1% change on something really stable). So perhaps it's a feature.

**Comment by letmaik** [comment]
It looks great, I like the arrow line and I think it's easy to interpret. I would call the scaling "issue" a feature.
Regarding good/bad semantics, yes, I think this is something for the config and very useful, if only to make scanning it visually half a second faster. Does it make sense to have a default of lower=better  so that you only override the inverted ones? I don't have a good intuition which direction is more common, but maybe by displaying the wrong color people are more encouraged to fix it, rather than if by default you have "meaningless" colors like now.

By the way, if you want to trigger CI without changing the code you can also create an empty commit.

**Comment by achamayou** [comment]
I’ve been wondering if we couldn’t just have some heuristics, and let users define regexps as overrides.
I would prefer to avoid single label spec because we generate them from a matrix and it’s incredibly tedious to spell each one out.

Anything that says “tx/s” or “throughput” is probably higher/better, whereas anything that ends in “ms”, “us” etc or contains “latency” likely isn’t.
Or maybe have a convention that “(unit)” is higher/better and “[unit]” lower/better? Or “unit^”?
I agree that if we require config we should use a clearly neutral colour (which can be the same for either direction, maybe a light gray).

**Comment by achamayou** [comment]
I read your comment wrongly, I see your point about displaying the wrong colour in absence of config. That sounds ok to me too, I suspect in general minimisation is more common (since timing, memory etc are straightforward sums whereas rates require division).

In that case, what about defaulting to lower is better, and flipping that if the metric name ends in “^”?

Ignore the bit about regexps, it’s misguided.

**Comment by letmaik** [comment]
Defaulting to lower is better sounds fine, but I don't like including metadata in the name. I thought about storing it in a separate db collection but it's all too complicated. Why are you dumping the regex idea?

**Comment by achamayou** [comment]
The thing I don’t like about storing them per metric isn’t the waste of space (which is insignificant, let’s face it), but the fact that it can be flipped and there’s no obvious way to handle that.

Sticking it in the label doesn’t have they problem because a new label is a new label and we know what to do with that.

User defined regexps are possibly ok, but having some preloaded ones just seems too opaque (there’s no easy way to explain how the direction works). Also there’s the problem of what to do if two contradictory regexps match...

A default to down and a switch is easy to explain/understand, it’s just unclear what the best way to implement the switch is.

**Comment by letmaik** [comment]
For regexp I wouldn't have preloaded ones, just user-defined, but I see your point, it's another indirection that complicates things. Let's do the label-based approach then, it's easy. So last character `^` and if it exists then it gets cut off (for the plot title) and means higher is better, right?
Example: `Throughput (tx/s)^`

**Comment by jumaffre** [comment]
Looks awesome! I wonder whether we should retain the old display and have it user-configurable in the `metrics.yml` file? At least for now, to get around the scalability issue. 
Also, it was not immediately clear to me what the dots represent on the graphs but I don't have a better idea right now.

**Comment by achamayou** [comment]
@letmaik, ok let’s try that and see how it goes. Having a separate collection for flags doesn’t sound like a problem, the more difficult question is I think how we populate it. I guess there are only two answer, either at the call site where we register the metric, or in the config.

The config sounds unattractive to me because it’s easy to get it out of sync or incomplete if the metrics are programmatically generated from a matrix etc.

The call site suffers from the potential issue of what to do if the flags change for the same label. One answer may be not to use the label as the id and to hash the label and the flags together. There’s a slight loss in readability on the DB side, but that’s meant to stay hidden.

I think that if we were only going to add that one flag, a label convention may be ok. If we want more, it’s clearly a dead end.

@jumaffre we definitely can retain both displays, I’m also not sure how to make the meaning of the points more obvious. Perhaps de-emphasise them by making them gray?
I think they give a better answer to “how stable is this number” than tweaking alpha or width on the bar though.

**Comment by ghost** [comment]
sparks@[315 aka 20200104.1](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=315&view=results) vs master ewma over 20 builds from [166](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=166&view=results) to [304](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=304&view=results)
images

**Comment by ghost** [comment]
sparks@[317 aka 20200104.3](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=317&view=results) vs master ewma over 20 builds from [166](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=166&view=results) to [304](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=304&view=results)
images

**Comment by ghost** [comment]
sparks@[319 aka 20200104.5](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=319&view=results) vs master ewma over 20 builds from [166](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=166&view=results) to [304](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=304&view=results)
images

**Comment by ghost** [comment]
sparks@[321 aka 20200104.7](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=321&view=results) vs master ewma over 20 builds from [166](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=166&view=results) to [304](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=304&view=results)
images

**Comment by ghost** [comment]
sparks@[322 aka 20200104.8](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=322&view=results) vs master ewma over 20 builds from [166](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=166&view=results) to [304](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=304&view=results)
images

**Comment by ghost** [comment]
sparks@[324 aka 20200104.10](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=324&view=results) vs master ewma over 20 builds from [166](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=166&view=results) to [304](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=304&view=results)
images

**Comment by ghost** [comment]
sparks@[325 aka 20200104.11](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=325&view=results) vs master ewma over 20 builds from [166](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=166&view=results) to [304](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=304&view=results)
images

**Comment by ghost** [comment]
sparks@[326 aka 20200104.12](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=326&view=results) vs master ewma over 20 builds from [166](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=166&view=results) to [304](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=304&view=results)
images

**Comment by ghost** [comment]
sparks@[327 aka 20200104.13](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=327&view=results) vs master ewma over 20 builds from [166](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=166&view=results) to [304](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=304&view=results)
images

**Comment by ghost** [comment]
sparks@[328 aka 20200104.14](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=328&view=results) vs master ewma over 20 builds from [166](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=166&view=results) to [304](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=304&view=results)
images

**Comment by ghost** [comment]
sparks@[329 aka 20200104.15](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=329&view=results) vs master ewma over 20 builds from [166](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=166&view=results) to [304](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=304&view=results)
images

**Comment by achamayou** [comment]
@jumaffre @letmaik ok, I think this is now wrapped up and ready for review. There's room for future improvements still, but I think it's useful enough to push out as an opt-in feature.

I had a look at:

1. Using a flag/separate collection. It's a bit of a pain, I cowardly decide to do this separately, and do the caret thing for now.
2. Plotting the branch-side series (as opposed to the last value). That's also less straightforward than I thought, because it can't meaningfully overlap with the historical data (otherwise, what does the x-scale even mean?). The right thing to do is probably to plot it to the right, but then it looks like a continuation of the same series, whereas it really is half ewma, half real values.
So I also cowardly claim this will be done later, once I've worked out a good way to do it.

**Comment by jumaffre** [comment]
Looks good to me! 
It seems that the accuracy metric has lost its historical values - I guess it's got nothing to do with the plotting? 
Also, should we strip the `" ^"` when displaying the name of the metrics and briefly update `README.md` according to the new changes?

**Review by jumaffre — APPROVED** [review]

File: cimetrics/env.py [hunk 1]
```diff
@@ -68,6 +76,8 @@ def github_token(self) -> str:
 
 
 class GitEnv(Env):
+    _target_branch = None
```
Comment (letmaik) [inline cimetrics/env.py:79]: This is a static class variable, but I think you meant to use an instance variable (AFAIK they need to go into `__init__`), even though here it works by accident.

File: cimetrics/plot.py [hunk 2]
```diff
@@ -175,8 +191,145 @@ def split(self, series):
...
+                tgt_raw.index.values[0],
+                tgt_raw.index.values[-span],
+                tgt_raw.index.values[-1],
+            ],
+            {"fontsize": 6},
+        )
+
+    plt.tight_layout()
+    plt.savefig(os.path.join(metrics_path, "diff.png"))
+
```
Comment (letmaik) [inline cimetrics/plot.py:325]: `plt.close(fig)`

File: cimetrics/plot.py [hunk 3]
```diff
@@ -246,3 +399,8 @@ def split(self, series):
     xticks = mtick.FormatStrFormatter(fmt)
     ax.xaxis.set_major_formatter(xticks)
     plt.savefig(os.path.join(metrics_path, "diff.png"))
+
+
+if __name__ == "__main__":
+    env = get_env()
+    (trend_view if env.view == "trend" else default_view)(env)
```
Comment (letmaik) [inline cimetrics/plot.py:410]: Shouldn't this be hidden inside cimetrics?

File: app/main.py [hunk 4]
```diff
@@ -22,11 +22,11 @@ def run_benchmark():
 results = run_benchmark()
 
 with cimetrics.upload.metrics() as m:
-    m.put("Throughput (tx/s)", results["throughput"])
+    m.put("Throughput (tx/s) ^", results["throughput"])
```
Comment (letmaik) [inline app/main.py:25]: The `^` should be removed and then the label trimmed for plotting, right?

File: cimetrics/env.py (hunk 1 above)
Comment (achamayou) [reply cimetrics/env.py:79]: This is deliberate, it's not an accident. Instance variables get initialised from static/class variable, it's a very common pattern to set attribute defaults in Python.

```
>>> class C:
...     a = 5
... 
>>> c1 = C()
>>> c1.a = 6
>>> c2 = C()
>>> c2.a
5
```

Comment (achamayou) [reply cimetrics/env.py:79]: See: https://rushter.com/blog/python-class-internals/

**Comment by ghost** [comment]
sparks@[330 aka 20200106.1](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=330&view=results) vs master ewma over 20 builds from [166](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=166&view=results) to [304](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=304&view=results)
images

File: cimetrics/plot.py (hunk 3 above)
Comment (achamayou) [reply cimetrics/plot.py:410]: How do you mean?

File: app/main.py (hunk 4 above)
Comment (achamayou) [reply app/main.py:25]: I don't know, I thought so at first, but then it seems like it's useful to have it to indicate that the expected direction is opposite to the default?

I'm almost seriously considering adding (📈👍) and (📉👍) next to each plot.

File: cimetrics/plot.py (hunk 3 above)
Comment (achamayou) [reply cimetrics/plot.py:410]: Oh, exposed as an API call as opposed to considering that the CLI is the interface?

Comment (letmaik) [reply cimetrics/plot.py:410]: It just seems weird that the end user has to inspect the env manually whereas all the rest of the env is dealt with by the library.

File: cimetrics/env.py (hunk 1 above)
Comment (letmaik) [reply cimetrics/env.py:79]: Didn't know that, ok, all fine then.

File: app/main.py (hunk 4 above)
Comment (letmaik) [reply app/main.py:25]: Hm, doesn't the arrow color in the graph indicate this already?

File: cimetrics/plot.py (hunk 3 above)
Comment (achamayou) [reply cimetrics/plot.py:410]: Ok, sure, I'll hide that behind a call. I just didn't expect any library use of plot.

File: app/main.py (hunk 4 above)
Comment (achamayou) [reply app/main.py:25]: Ok, I'll hide it then!

**Comment by ghost** [comment]
sparks@[331 aka 20200106.2](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=331&view=results) vs master ewma over 20 builds from [166](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=166&view=results) to [304](https://dev.azure.com/jumaffre/metrics-devops/_build/results?buildId=304&view=results)
images

**Pull request merged**